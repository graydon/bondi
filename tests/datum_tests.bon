(*** datum tests *) 

(* 
%open "../tests/datum_tests.bon";; 
*) 

"datum values";;

3;;
3.3;;
'a';;
"a";;
Un;;


(* int_ops_bss *) 

3 plusint 4 eqcons 7;;
5 minusint 3 eqcons 2;;
3 timesint 4 eqcons 12;
9 divideint 4 eqcons 2;;
7 modint 4 eqcons 3;;

(* int_ops1 *) 

negateint 3 eqcons (-3);;

(*
randomint 10000;;
*)
"float_ops1";;

acos 0.5;;
asin 0.5;;
atan 0.5;;
ceil 0.5;;
cos 0.5;;
cosh 0.5;;
exponential 0.5;;
fabs 0.5;;
floor 0.5;;
log 0.5;;
log10 0.5;;
negatefloat 0.5;;
(*
randomfloat 0.5;;
*)
sin 0.5;;
sinh 0.5;;
sqrt 0.5;;
tan 0.5;;
tanh 0.5;;

(* float_ops2  

  "atan2";
  "fmod";
  "pow"
*)

(* float_ops_bss *) 

4.4 plusfloat 3.3;;
4.4 minusfloat 3.3;;
4.4 timesfloat 3.3;;
4.4 dividefloat 3.3;;
4.4 atan2 3.3;;
4.4 fmod 3.3;;
4.4 pow 3.3;;


(* int_rels *) 

3 lessthanint 4;;
3 lessthanorequalint 4;;
5 greaterthanint 4;;
4 greaterthanorequalint 4;;
 

(* float_rels *) 

3.3 lessthanfloat 4.4;;
3.3 lessthanorequalfloat 4.4;;
not (3.3 greaterthanfloat 4.4);;
4.4 greaterthanorequalfloat 4.4;;
 

(* char_rels *)  

'a' lessthanchar'b' ;;
'a' lessthanorequalchar'b' ;;
'b' greaterthanchar 'a' ;;
not ('a' greaterthanorequalchar 'b') ;;

(* other ops, mainly string ops *) 
 
int2float 3;;
truncate 3.3;;
printchar 'c';;
printstring "abc";;
"abc" concat "def";;
"abc" getcharstring 2;;
4 makestring 'f';;
"abbc" indexstring 'b';;
"abbc" rindexstring 'b';;
"abbc" containsstring 'b';;
copystring "abc";;
escapedstring "abc";;
lengthstring "abc";;
uppercasestring "aBc";;
substring "abcd" 1 3;;
indexfromstring "abcd" 1 'b' ;;
containsfromstring "abcd" 1 'b' ;;

(* printing chars and strings  
	printchar and printstring do not quote.
	prim2string adds single or double quotes. 
	the generic print calls prim2string.
	To display text use printchar and printstring.
	To discuss text (as data) use print. 	
		*) 

printchar 'a';;
printstring (prim2string 'a');;
prim2string 'a';;
'a';;

printstring "abc";;
printstring (prim2string "abc");;
prim2string "abc";;
"abc";;

printchar '\n';;
printchar '\t';;
printchar 'a';;


printstring "At present, newlines and other whitespace" ; 
printchar '\n' ; 
printstring "are only accessible as characters."  ; 
printchar '\n';
printstring "To change this, modify the lexer for strings." ;
printchar '\n'
;;

(* wildcards *) 

_Int;;
_Float;;
_Char;;
_String;;

%show nomatch;;

negateint _Int == (Exception "nomatch");;
lengthstring _String == (Exception "nomatch");;

%hide nomatch;;
