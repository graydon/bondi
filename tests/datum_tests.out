(* Begin sourcing "./prelude/standard_prelude.bon"... *)
(* Begin sourcing "./prelude/printing.bon"... *)
^: String -> String -> String
toString: a -> String
print: a -> Unit
println: a -> Unit
(* Finished sourcing "./prelude/printing.bon" *)
(* Begin sourcing "./prelude/imperatives_core.bon"... *)
forall: Int -> Int -> (Int -> a) -> Unit
iter: (all a.a -> Unit) -> b -> Unit
clone: a -> a
(* Finished sourcing "./prelude/imperatives_core.bon" *)
(* Begin sourcing "./prelude/structure_poly.bon"... *)
Pair: a -> b -> a * b
fromTuple: a -> String
it: Unit
fst: a * b -> a
snd: a * b -> b
Inl: a -> a + b
Inr: b -> a + b
Evr: a -> Konstant a b
Ths: a -> Identity a
ParamPair: a b -> c b -> ParamProduct a c b
Ok: a b -> Okay a b
Nest: a (b c) -> Nested a b c
Tag: (a -> b) -> a c -> Represent a b c
deconstruct: a b -> (all c.Represent c a b -> d) -> d
reconstruct: Represent a b c -> b c
map: (a -> b) -> c a -> c b
Nil: List a
Cons: a -> List a -> List a
toString: b -> String
toString: c -> String
Some: a -> Maybe a
None: Maybe a
foldleft: (a -> b -> a) -> a -> c b -> a
foldright: (a -> b -> b) -> c a -> b -> b
zipwith: (a * b -> c) -> d a * e b -> d c
(* Finished sourcing "./prelude/structure_poly.bon" *)
(* Begin sourcing "./prelude/datum2generic.bon"... *)
&&: Bool -> Bool -> Bool
not: Bool -> Bool
||: Bool -> Bool -> Bool
equal: a * b -> Bool
==: a -> a -> Bool
!=: a -> a -> Bool
plus: a * b -> a
+: a -> a -> a
shortplus: a * b -> a
minus: a * b -> a
-: a -> a -> a
times: a * b -> a
*: a -> a -> a
divide: a * b -> a
/: a -> a -> a
negate: a -> a
lessthan: a * b -> Bool
<: a -> a -> Bool
lessthanorequal: a * b -> Bool
<=: a -> a -> Bool
greaterthan: a * b -> Bool
>: a -> a -> Bool
greaterthanorequal: a * b -> Bool
>=: a -> a -> Bool
(* Finished sourcing "./prelude/datum2generic.bon" *)
(* Begin sourcing "./prelude/dynamic2static.bon"... *)
dynamic_case_for_internal_use_only = ...
(* Finished sourcing "./prelude/dynamic2static.bon" *)
(* Begin sourcing "./prelude/queries.bon"... *)
apply2all: (all a.a -> a) -> b -> b
append: List a -> List a -> List a
some_append: Maybe a -> List a -> List a
select: (all a.a -> Maybe b) -> c -> List b
is: lin a -> b -> Maybe a
(* Finished sourcing "./prelude/queries.bon" *)
(* Finished sourcing "./prelude/standard_prelude.bon" *)
(* Begin sourcing "./datum_tests.bon"... *)
it: String
it = "datum values"
it: Int
it = 3
it: Float
it = 3.3
it: Char
it = 'a'
it: String
it = "a"
it: Unit
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: String
it = "float_ops1"
it: Float
it = 1.0471975512
it: Float
it = 0.523598775598
it: Float
it = 0.463647609001
it: Float
it = 1.
it: Float
it = 0.87758256189
it: Float
it = 1.12762596521
it: Float
it = 1.6487212707
it: Float
it = 0.5
it: Float
it = 0.
it: Float
it = -0.69314718056
it: Float
it = -0.301029995664
it: Float
it = -0.5
it: Float
it = 0.479425538604
it: Float
it = 0.521095305494
it: Float
it = 0.707106781187
it: Float
it = 0.546302489844
it: Float
it = 0.46211715726
it: Float
it = 7.7
it: Float
it = 1.1
it: Float
it = 14.52
it: Float
it = 1.33333333333
it: Float
it = 0.927295218002
it: Float
it = 1.1
it: Float
it = 132.859872828
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Float
it = 3.
it: Int
it = 3
it: Unit
cit: Unit
abcit: String
it = "abcdef"
it: Char
it = 'c'
it: String
it = "ffff"
it: Int
it = 1
it: Int
it = 2
it: Bool
it = True
it: String
it = "abc"
it: String
it = "abc"
it: Int
it = 3
it: String
it = "ABC"
it: String
it = "bcd"
it: Int
it = 1
it: Bool
it = True
it: Unit
ait: Unit
'a'it: String
it = "'a'"
it: Char
it = 'a'
it: Unit
abcit: Unit
"abc"it: String
it = ""abc""
it: String
it = "abc"
it: Unit

it: Unit
	it: Unit
ait: Unit
At present, newlines and other whitespace
are only accessible as characters.
To change this, modify the lexer for strings.
it: Int
it = _Int
it: Float
it = _Float
it: Char
it = _Char
it: String
it = _String
it: Bool
it = True
it: Bool
it = True
(* Finished sourcing "./datum_tests.bon" *)
